File: docs/algorithms/code/algorithms/isValid.ts
function isValid(s: string): boolean {
  const stack: string[] = [];

  const map: Record<string, any> = {
    ")": "(",
    "}": "{",
    "]": "[",
  };

  for (let i = 0; i < s.length; i++) {
    if (stack.length && stack[stack.length - 1] === map[s[i]]) {
      stack.pop();
    } else {
      stack.push(s[i]);
    }
  }

  return Boolean(stack.length === 0);
}

console.log(isValid("()"));

File: docs/algorithms/code/algorithms/isValid.go
// 思路：使用栈存储，遇到左括号入栈，遇到右括号出栈，判断是否匹配
// 复杂度分析：时间复杂度 O(n)，空间复杂度 O(n)
func isValid(s string) bool {
	hash := map[string]string{
		")": "(",
		"]": "[",
		"}": "{",
	}

	stack := []string{}

	for _, val := range s {
		if val == '(' || val == '[' || val == '{' {
			stack = append(stack, string(val))
		} else {
			if len(stack) == 0 {
				return false
			}
			v := stack[len(stack)-1]
			if hash[string(val)] == v {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}

	return len(stack) == 0
}